// @aiteq/messenger-bot/bin/fb-api/messenger-codes.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Graph = require("./graph-api");
class Api extends Graph.Api {
    constructor(accessToken) {
        super(accessToken, Graph.Endpoint.MESSENGER_CODES);
        this.accessToken = accessToken;
    }
    generateCode(size, ref) {
        return __awaiter(this, void 0, void 0, function* () {
            const req = {
                type: Type.STANDARD,
                image_size: size || 1000
            };
            ref && (req.data = { ref });
            try {
                return (yield this.sendRequest(req)).uri;
            }
            catch (error) {
                return Promise.reject(error);
            }
        });
    }
}
exports.Api = Api;
var Type;
(function (Type) {
    Type["STANDARD"] = "standard";
})(Type = exports.Type || (exports.Type = {}));
