// @aiteq/messenger-bot/bin/fb-api/webhook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var PostbackSource;
(function (PostbackSource) {
    PostbackSource["GET_STARTED_BUTTON"] = "get-started-button";
    PostbackSource["POSTBACK_BUTTON"] = "postback-button";
    PostbackSource["PERSISTENT_MENU"] = "persistent-menu";
})(PostbackSource = exports.PostbackSource || (exports.PostbackSource = {}));
var AttachmentType;
(function (AttachmentType) {
    AttachmentType["IMAGE"] = "image";
    AttachmentType["AUDIO"] = "audio";
    AttachmentType["VIDEO"] = "video";
    AttachmentType["FILE"] = "file";
    AttachmentType["LOCATION"] = "location";
    AttachmentType["FALLBACK"] = "fallback";
})(AttachmentType = exports.AttachmentType || (exports.AttachmentType = {}));
var ReferralSource;
(function (ReferralSource) {
    ReferralSource["SHORTLINK"] = "SHORTLINK";
    ReferralSource["ADS"] = "ADS";
    ReferralSource["MESSENGER_CODE"] = "MESSENGER_CODE";
    ReferralSource["DISCOVER_TAB"] = "DISCOVER_TAB";
})(ReferralSource = exports.ReferralSource || (exports.ReferralSource = {}));
var Event;
(function (Event) {
    Event["TEXT_MESSAGE"] = "message";
    Event["MESSAGE_DELIVERED"] = "delivery";
    Event["MESSAGE_READ"] = "read";
    Event["MESSAGE_ECHO"] = "echo";
    Event["POSTBACK"] = "postback";
    Event["ATTACHMENT"] = "attachment";
    Event["ATTACHMENT_IMAGE"] = "attachment:image";
    Event["ATTACHMENT_AUDIO"] = "attachment:audio";
    Event["ATTACHMENT_VIDEO"] = "attachment:video";
    Event["ATTACHMENT_FILE"] = "attachment:file";
    Event["ATTACHMENT_FALLBACK"] = "attachment:fallback";
    Event["ATTACHMENT_LOCATION"] = "attachment:location";
    Event["GET_STARTED_BUTTON"] = "postback:get";
    Event["POSTBACK_BUTTON"] = "postback:get-started-button";
    Event["PERSISTENT_MENU"] = "postback:persistent-menu";
    Event["TEXT_QUICK_REPLY"] = "qr:text";
    Event["LOCATION_QUICK_REPLY"] = "qr:location";
})(Event = exports.Event || (exports.Event = {}));
