// @aiteq/messenger-bot/bin/fb-api-helpers/default-action-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fb_api_1 = require("../fb-api");
const builder_1 = require("./builder");
class DefaultActionBuilder extends builder_1.Builder {
    constructor(url) {
        super();
        this.action = {
            type: fb_api_1.Send.ButtonType.WEB_URL,
            url
        };
    }
    setWebviewHeightRatio(webviewHeightRatio) {
        this.action.webview_height_ratio = webviewHeightRatio;
        return this;
    }
    setMessengerExtensions(messengerExtensions) {
        this.action.messenger_extensions = messengerExtensions;
        return this;
    }
    setFallbackUrl(fallbackUrl) {
        this.action.fallback_url = fallbackUrl;
        return this;
    }
    setShowWebviewShareButton(visible) {
        this.action.webview_share_button = visible === true ? fb_api_1.Webview.ShareButton.SHOW : fb_api_1.Webview.ShareButton.HIDE;
        return this;
    }
    build() {
        return this.action;
    }
}
exports.DefaultActionBuilder = DefaultActionBuilder;
