// @aiteq/messenger-bot/bin/fb-api-helpers/generic-template-message-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fb_api_1 = require("../fb-api");
const template_message_builder_1 = require("./template-message-builder");
class GenericTemplateMessageBuilder extends template_message_builder_1.TemplateMessageBuilder {
    constructor() {
        super();
        this.template = {
            template_type: fb_api_1.Send.TemplateType.GENERIC,
            elements: new Array()
        };
    }
    setImageAspectRatio(imageAspectRatio) {
        this.template.image_aspect_ratio = imageAspectRatio;
        return this;
    }
    setSherable(sherable) {
        this.template.sherable = sherable;
        return this;
    }
    addElement(elementBuilder) {
        if (this.template.elements.length === 10) {
            throw new Error("couldn't add next Element to Generic Tepmplate message (only 10 elements is allowed)");
        }
        this.template.elements.push(elementBuilder.build());
        return this;
    }
}
exports.GenericTemplateMessageBuilder = GenericTemplateMessageBuilder;
