// @aiteq/messenger-bot/bin/fb-api-helpers/list-template-message-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fb_api_1 = require("../fb-api");
const button_holding_template_message_builder_1 = require("./button-holding-template-message-builder");
class ListTemplateMessageBuilder extends button_holding_template_message_builder_1.ButtonHoldingTemplateMessageBuilder {
    constructor() {
        super();
        this.template = {
            template_type: fb_api_1.Send.TemplateType.LIST,
            elements: new Array()
        };
    }
    setTopElementStyle(topElementStyle) {
        this.template.top_element_style = topElementStyle;
        return this;
    }
    addElement(elementBuilder) {
        if (this.template.elements.length === 4) {
            throw new Error("couldn't add next Element to List Tepmplate message (only 2-4 elements is allowed)");
        }
        this.template.elements.push(elementBuilder.build());
        return this;
    }
    addButton(button) {
        this.template.buttons = [button];
    }
}
exports.ListTemplateMessageBuilder = ListTemplateMessageBuilder;
