// @aiteq/messenger-bot/bin/fb-api-helpers/message-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fb_api_1 = require("../fb-api");
const builder_1 = require("./builder");
class MessageBuilder extends builder_1.Builder {
    addQuickReply(title, id, data, imageUrl) {
        this.message.quick_replies = this.message.quick_replies || new Array();
        const payload = { id };
        data && (payload.data = typeof data === "string" ? data : JSON.stringify(data));
        const qr = {
            content_type: fb_api_1.Send.ContentType.TEXT,
            title,
            payload: JSON.stringify(payload)
        };
        imageUrl && (qr.image_url = imageUrl);
        this.message.quick_replies.push(qr);
        return this;
    }
    addLocationQuickReply() {
        this.message.quick_replies = this.message.quick_replies || new Array();
        this.message.quick_replies.push({ content_type: fb_api_1.Send.ContentType.LOCATION });
        return this;
    }
    build() {
        return this.message;
    }
}
exports.MessageBuilder = MessageBuilder;
